File: backend/app/api/v1/transactions.py
from typing import Optional
from fastapi import APIRouter, Depends, Query, Response, status
from app.core.deps import require_admin
from app.schemas.transaction import TransactionCreateRequest
from app.services.receipt_service import receipt_service
from app.services.transaction_service import transaction_service
from app.utils.responses import success_response

router = APIRouter(prefix="/admin/transactions", tags=["Transaksi Bank Sampah (Admin)"])


@router.get("")
def list_transactions(
    start_date: Optional[str] = Query(None, description="Tanggal awal (YYYY-MM-DD)"),
    end_date: Optional[str] = Query(None, description="Tanggal akhir (YYYY-MM-DD)"),
    type: Optional[str] = Query(None, description="SETOR atau TARIK"),
    category_id: Optional[str] = Query(None, description="ID Kategori sampah"),
    nasabah_id: Optional[str] = Query(None, description="ID Nasabah"),
    search: Optional[str] = Query(None, description="Pencarian nomor transaksi atau nama/NIK nasabah"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    admin_user: dict = Depends(require_admin)
):
    """Daftar semua transaksi dengan filter dan pagination."""
    items, total_items = transaction_service.list_transactions(
        nasabah_id=nasabah_id,
        start_date=start_date,
        end_date=end_date,
        type_filter=type,
        category_id=category_id,
        search=search,
        page=page,
        page_size=page_size
    )
    total_pages = (total_items + page_size - 1) // page_size if total_items > 0 else 0
    return success_response(
        data={
            "items": items,
            "pagination": {
                "page": page,
                "page_size": page_size,
                "total_items": total_items,
                "total_pages": total_pages,
            }
        }
    )


@router.post("", status_code=status.HTTP_201_CREATED)
def create_transaction(
    data: TransactionCreateRequest, admin_user: dict = Depends(require_admin)
):
    """Mencatat transaksi baru (SETOR sampah atau TARIK tunai)."""
    tx = transaction_service.create_transaction(data, creator_id=admin_user["id"])
    msg = "Transaksi setor berhasil disimpan." if data.type == "SETOR" else "Transaksi tarik tunai berhasil disimpan."
    return success_response(data=tx, message=msg, status_code=status.HTTP_201_CREATED)


@router.get("/{transaction_id}")
def get_transaction_detail(transaction_id: str, admin_user: dict = Depends(require_admin)):
    """Detail lengkap transaksi."""
    tx = transaction_service.get_transaction_by_id(transaction_id)
    return success_response(data=tx)


@router.get("/{transaction_id}/receipt")
def get_transaction_receipt(
    transaction_id: str,
    format: Optional[str] = Query("json", description="Format output: json atau pdf"),
    admin_user: dict = Depends(require_admin)
):
    """Mendapatkan bukti transaksi dalam format JSON atau file PDF (Admin)."""
    if format == "pdf":
        pdf_bytes = receipt_service.generate_receipt_pdf(transaction_id, admin_user)
        filename = f"bukti-transaksi-{transaction_id[:8]}.pdf"
        return Response(
            content=pdf_bytes,
            media_type="application/pdf",
            headers={"Content-Disposition": f"inline; filename={filename}"}
        )
    
    receipt_data = receipt_service.get_receipt_data(transaction_id, admin_user)
    return success_response(data=receipt_data)

File: backend/src/routes/transactions.ts
import { Hono } from 'hono';
import { AppEnv } from '../types/index.js';
import { transactionService } from '../services/transactionService.js';
import { receiptService } from '../services/receiptService.js';
import { authMiddleware } from '../middleware/auth.js';
import { requireAdmin } from '../middleware/roleGuard.js';
import { successResponse } from '../utils/response.js';
import { validateJson } from '../utils/validator.js';
import { transactionCreateSchema } from '../schemas/transaction.schema.js';

export const transactionsRouter = new Hono<AppEnv>();

// All routes require ADMIN
transactionsRouter.use('*', authMiddleware, requireAdmin);

// GET /api/v1/admin/transactions
transactionsRouter.get('/', async (c) => {
  const query = c.req.query();
  const result = await transactionService.listTransactions({
    nasabah_id: query.nasabah_id,
    start_date: query.start_date,
    end_date: query.end_date,
    type: query.type,
    category_id: query.category_id,
    search: query.search,
    page: query.page ? parseInt(query.page, 10) : 1,
    page_size: query.page_size ? parseInt(query.page_size, 10) : 20,
  });
  return successResponse(c, result, 'Daftar transaksi berhasil dimuat.');
});

// POST /api/v1/admin/transactions
transactionsRouter.post('/', validateJson(transactionCreateSchema), async (c) => {
  const user = c.get('user');
  const data = c.req.valid('json' as any);
  const result = await transactionService.createTransaction(data, user.id);
  const actionName = data.type === 'SETOR' ? 'Setor sampah' : 'Tarik tunai';
  return successResponse(c, result, `${actionName} berhasil dicatat.`, 201);
});

// GET /api/v1/admin/transactions/:id
transactionsRouter.get('/:id', async (c) => {
  const id = c.req.param('id');
  const result = await transactionService.getTransactionById(id);
  return successResponse(c, result, 'Detail transaksi berhasil dimuat.');
});

// GET /api/v1/admin/transactions/:id/receipt
transactionsRouter.get('/:id/receipt', async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const format = c.req.query('format');

  if (format === 'pdf') {
    const pdfBuffer = await receiptService.generateReceiptPdf(id, user);
    c.header('Content-Type', 'application/pdf');
    c.header('Content-Disposition', `inline; filename="bukti-transaksi-${id.slice(0, 8)}.pdf"`);
    return c.body(new Uint8Array(pdfBuffer));
  }

  const receiptData = await receiptService.getReceiptData(id, user);
  return successResponse(c, receiptData);
});
